what4/src/What4/Utils/BVDomain.hs

```diff
@@ -1,18 +1,503 @@
+-- @testBit@, the least significant bit has index @0@.
+range :: NatRepr w -> Integer -> Integer -> BVDomain w
+range w al ah = interval mask al (ah - al)
+  where mask = maxUnsigned w
+
+-- | Unsafe constructor for internal use only. Caller must ensure that
+-- @mask = maxUnsigned w@.
+interval :: Integer -> Integer -> Integer -> BVDomain w
+interval mask al aw =
+  if aw >= mask then BVDAny mask else BVDInterval mask (al .&. mask) aw
+
+-- | Create an abstract domain from an ascending list of elements.
+-- The elements are assumed to be distinct.
+fromAscEltList :: (1 <= w) => NatRepr w -> [Integer] -> BVDomain w
+fromAscEltList w [] = singleton w 0
+fromAscEltList w [x] = singleton w x
+fromAscEltList w (x0 : x1 : xs) = go (x0, x0) (x1, x1) xs
+  where
+    -- Invariant: the gap between @b@ and @c@ is the biggest we've
+    -- seen between adjacent values so far.
+    go (a, b) (c, d) [] = union (range w a b) (range w c d)
+    go (a, b) (c, d) (e : rest)
+      | e - d > c - b = go (a, d) (e, e) rest
+      | otherwise     = go (a, b) (c, e) rest
+
+-- | Return union of two domains.
+union :: (1 <= w) => BVDomain w -> BVDomain w -> BVDomain w
+union a b =
+  case a of
+    BVDAny _ -> a
+    BVDInterval _ al aw ->
+      case b of
+        BVDAny _ -> b
+        BVDInterval mask bl bw ->
+          interval mask cl (ch - cl)
+          where
+            size = mask + 1
+            ac = 2 * al + aw
+            bc = 2 * bl + bw
+            al' = if ac + mask < bc then al + size else al
+            bl' = if bc + mask < ac then bl + size else bl
+            ah' = al' + aw
+            bh' = bl' + bw
+            cl = min al' bl'
+            ch = max ah' bh'
+
+-- | @concat a y@ returns domain where each element in @a@ has been
+-- concatenated with an element in @y@.  The most-significant bits
+-- are @a@, and the least significant bits are @y@.
+concat :: NatRepr u -> BVDomain u -> NatRepr v -> BVDomain v -> BVDomain (u + v)
+concat u a v b =
+  case a of
+    BVDAny _ -> BVDAny mask
+    BVDInterval _ al aw -> interval mask (cat al bl) (cat aw bw)
+  where
+    cat i j = i `shiftL` widthVal v + j
+    mask = maxUnsigned (addNat u v)
+    (bl, bh) = ubounds b
+    bw = bh - bl
+
+-- | @shrink i a@ drops the @i@ least significant bits from @a@.
+shrink ::
+  NatRepr i ->
+  BVDomain (i + n) -> BVDomain n
+shrink i a =
+  case a of
+    BVDAny mask -> BVDAny (shr mask)
+    BVDInterval mask al aw ->
+      interval (shr mask) bl (bh - bl)
+      where
+        bl = shr al
+        bh = shr (al + aw)
+  where
+    shr x = x `shiftR` widthVal i
+
+-- | @trunc n d@ selects the @n@ least significant bits from @d@.
+trunc ::
+  (n <= w) =>
+  NatRepr n ->
+  BVDomain w -> BVDomain n
+trunc n a =
+  case a of
+    BVDAny _ -> BVDAny mask
+    BVDInterval _ al aw -> interval mask al aw
+  where
+    mask = maxUnsigned n
+
+-- | @select i n a@ selects @n@ bits starting from index @i@ from @a@.
+select ::
```

Comment:
I like that this is easy to understand, but can we do the computation with fewer intermediate values if we combine the operations?

Reply:
I think if they were combined, you'd still have to compute all the same intermediate values. Doing a select on an Integer in Haskell still requires separate mask and shift operations.